# Coeficientes De Relación

In [1]:
#Importando Librerias a utilizar
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import spearmanr

In [2]:
#Leyendo Dataset
datos= pd.read_csv("data.csv", index_col=False)

En este caso elimino las columnas que no seran necesarias ya que buscando una correlacion entre ciertas variables de forma especifica.

In [3]:
#Eliminando Columnas No Necesarias
datos.drop(columns=['participant','age:1', 'gender:1', 'noticias:1', 'cuarentena_1:1', 'cuarentena_2:1', 
                    'cuarentena_3:1','proximo:1', 'country', 'TIME_start', 'TIME_end', 'TIME_total',
                    'previo:1', 'previo:2', 'previo:3', 'previo:4'],inplace=True)
#Eliminando valores faltantes
datos.dropna(inplace=True)
#datos.astype(int)
#Agregando promedio de termino del test IGT en segundos
datos["igt:meantime"] = [473.93,709.29,1783.56,1177.78,1273.17,655.02,1295.9,1073.57,1084.94,425.82,
                     1032.39,822.43,936.39,1047.82,1667.18,467.68,389.36,1001.06,1842.02,1998.32,
                     2127.27,2581.46,1418.62,916.47]
#Agregando Valor final con el que terminaron (Dinero)
datos["igt:earn"] = [-500,3200,4000,2100,1950,1850,2300,1050,2450,4250,
                     1650,1950,2350,1900,650,3200,3100,1450,1800,50,
                     3350,2200,1200,1350]
#Reseteando Index
datos.reset_index(inplace=True, drop=True)
#Eliminando ultima columna no necesaria
datos.drop(columns="igt:1", inplace=True)
#Dividiento el dataframe para evitar convertir cada columna
datos2 = datos[["igt:meantime","igt:earn"]]
#Conviertiendo a tipo Entero
datos = datos[['ndc:1', 'ndc:2', 'ndc:3', 'ndc:4', 'ndc:5', 'ndc:6', 'ndc:7', 'ndc:8',
       'ndc:9', 'ndc:10', 'ndc:11', 'ndc:12', 'ndc:13', 'ndc:14', 'ndc:15',
       'ndc:16', 'ndc:17', 'ndc:18', 'beck:1', 'beck:2', 'beck:3', 'beck:4',
       'beck:5', 'beck:6', 'beck:7', 'beck:8', 'beck:9', 'beck:10', 'beck:11',
       'beck:12', 'beck:13', 'beck:14', 'beck:15', 'beck:16', 'beck:17',
       'beck:18', 'beck:19', 'beck:20', 'beck:21']].astype(int)
#Uniendo resultados otra vez
result = pd.concat([datos, datos2], axis=1, join='inner')

In [4]:
result.head()

ndc:1  ndc:2  ndc:3  ndc:4  ndc:5  ndc:6  ndc:7  ndc:8  ndc:9  ndc:10  ...  \
0      1      1      1      1      1      1      1      1      1       1  ...   
1      2      4      2      2      2      3      2      2      3       4  ...   
2      4      4      1      2      2      5      1      2      3       3  ...   
3      2      4      4      4      1      2      1      4      5       2  ...   
4      4      3      2      1      1      5      1      3      1       4  ...   

   beck:14  beck:15  beck:16  beck:17  beck:18  beck:19  beck:20  beck:21  \
0        0        0        0        0        0        0        0        0   
1        1        1        1        1        1        1        1        2   
2        0        3        0        0        3        0        0        0   
3        0        1        0        1        0        0        0        0   
4        0        0        0        0        1        0        0        1   

   igt:meantime  igt:earn  
0        473.93      -500  
1        709.29      3200  
2       1783.56      4000  
3       1177.78      2100  
4       1273.17      1950  

[5 rows x 41 columns]

## Busqueda De Correlación

Para que exista un relacion entre dos variables se realizan pruebas de hipotesis asumiento un valor de P-Valor= 0.05 contra el valor P-Valor obtenido de cada una de las variables que pueden tener relacion.

Aqui agrego algunos sitios webs y documentos que me sirvieron como una referencia. Si existe algun error, favor de decirme con total libertad.

* https://www.scribbr.com/methodology/types-of-variables/
* https://statistics.laerd.com/spss-tutorials/spearmans-rank-order-correlation-using-spss-statistics.ph
* https://statistics.laerd.com/spss-tutorials/pearsons-product-moment-correlation-using-spss-statistics.php
* https://www.statisticshowto.com/ranked-variable/
* https://businessjargons.com/spearmans-rank-correlation-coefficient.html
* https://medium.com/@outside2SDs/an-overview-of-correlation-measures-between-categorical-and-continuous-variables-4c7f85610365
*https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.spearmanr.html

Cabe destacar que en este caso al ser variables de tipo Cualitativas Ordinales el coeficiente de Pearson no puede ser utizado.
De la misma forma esta es una forma de buscar las correlaciones, sin embargo cabe destacar que otras mejores aproximaciones podrian haber sido realizar unas pruebas con Chi Square o una anova.
Dicho proceso se realizara en pruebas posteriores.

### Tiempo De Respuesta Vs Respuedas del NDC
El tiempo esta medido en segundos.

In [5]:
def timevsndc(lista, p):
    """
    Esta funcion recibe como parametro el numero de la pregunta de Beck Para evaluar con las preguntas de ndc (que son 18 en total)
    Input: String o Entero para seleccionar la pregunta o variable
    Output: Retorna una lista con las variables que tienen cierta relacion en base a una significancia estadistica.
    Igual Imprime cada una de ellas. 
    """
    #Lista Vacia    
    var_column = "igt:meantime"
    #Rango por la cantidad total de preguntas que le pertenecen al NDC
    for j in range(1,19): 
        #Valor que toma por variable
        ndc = "ndc:"+str(j)
        #Coeficiente de Pearson y P-valor entre las variables
        r, p_value = spearmanr(result[var_column], result[ndc])
        #print(r,p_value)
        if p_value <=p:
            string = f"El tiempo promedio con la pregunta #{str(j)} del ndc con un P-valor de: {p_value} y Coeficiente De Relaccion de: {r}" 
            lista.append(string)
            #lista2.append(r,p_value)
    return lista#,lista2

### P Valor < 0.05

In [10]:
lista=[]
lista = timevsndc(lista, p=0.05)
print(lista)

[]


### P Valor < 0.01

In [9]:
lista=[]
lista = timevsndc(lista, p=0.01)
print(lista)

[]


### P Valor < 0.001

In [11]:
lista=[]
lista = timevsndc(lista, p=0.001)
print(lista)

[]


### Tiempo De Respuesta Vs Respuedas del BECK
El tiempo esta medido en segundos.

In [14]:
def timevsbeck(lista, p):
    """
    Esta funcion recibe como parametro el numero de la pregunta de Beck Para evaluar con las preguntas de ndc (que son 18 en total)
    Input: String o Entero para seleccionar la pregunta o variable
    Output: Retorna una lista con las variables que tienen cierta relacion en base a una significancia estadistica.
    Igual Imprime cada una de ellas. 
    """
    #Lista Vacia    
    var_column = "igt:meantime"
    #Rango por la cantidad total de preguntas que le pertenecen al NDC
    for j in range(1,21): 
        #Valor que toma por variable
        ndc = "beck:"+str(j)
        #Coeficiente de Pearson y P-valor entre las variables
        r, p_value = spearmanr(result[var_column], result[ndc])
        #print(r,p_value)
        if p_value <=p:
            string = f"El tiempo promedio con la pregunta #{str(j)} del ndc con un P-valor de: {p_value} y Coeficiente De Relaccion de: {r}" 
            lista.append(string)
            #lista2.append(r,p_value)
    return lista#,lista2

### P Valor < 0.05

In [16]:
lista1=[]
lista1 = timevsbeck(lista1,p=0.05)
for element in lista1:
    print(element)

El tiempo promedio con la pregunta #1 del ndc con un P-valor de: 0.03232148859588864 y Coeficiente De Relaccion de: 0.43794581869247995
El tiempo promedio con la pregunta #9 del ndc con un P-valor de: 0.036519567387405166 y Coeficiente De Relaccion de: -0.4288614750612719


### P Valor < 0.01

In [17]:
lista1=[]
lista1 = timevsbeck(lista1,p=0.01)
for element in lista1:
    print(element)

### P Valor < 0.001

In [18]:
lista1=[]
lista1 = timevsbeck(lista1,p=0.01)
for element in lista1:
    print(element)

### Dinero Total Con el que termino el participante Vs Respuedas del NDC
Es comparado con la cantidad de dinero total con la que el participante termino al momento de finalizar la prueba del IGT

In [19]:
def igtevsndc(lista, p):
    """
    Esta funcion recibe como parametro el numero de la pregunta de Beck Para evaluar con las preguntas de ndc (que son 18 en total)
    Input: String o Entero para seleccionar la pregunta o variable
    Output: Retorna una lista con las variables que tienen cierta relacion en base a una significancia estadistica.
    Igual Imprime cada una de ellas. 
    """
    #Lista Vacia    
    var_column = "igt:earn"
    #Rango por la cantidad total de preguntas que le pertenecen al NDC
    for j in range(1,19): 
        #Valor que toma por variable
        ndc = "ndc:"+str(j)
        #Coeficiente de Pearson y P-valor entre las variables
        r, p_value = spearmanr(result[var_column], result[ndc])
        #print(r,p_value)
        if p_value <=p:
            string = f"El tiempo promedio con la pregunta #{str(j)} del ndc con un P-valor de: {p_value} y Coeficiente De Relaccion de: {r}" 
            lista.append(string)
            #lista2.append(r,p_value)
    return lista#,lista2

### P Valor < 0.05

In [20]:
lista2= []
lista2 = igtevsndc(lista2,p=0.05)
for element in lista2:
    print(element)

### P Valor < 0.001

In [21]:
lista2= []
lista2 = igtevsndc(lista2,p=0.01)
for element in lista2:
    print(element)

### P Valor < 0.001

In [22]:
lista2= []
lista2 = igtevsndc(lista2,p=0.001)
for element in lista2:
    print(element)

### Dinero Total Con el que termino el participante Vs Respuedas del BECK
Es comparado con la cantidad de dinero total con la que el participante termino al momento de finalizar la prueba del IGT

In [27]:
def igtevsbeck(lista, p):
    """
    Esta funcion recibe como parametro el numero de la pregunta de Beck Para evaluar con las preguntas de ndc (que son 18 en total)
    Input: String o Entero para seleccionar la pregunta o variable
    Output: Retorna una lista con las variables que tienen cierta relacion en base a una significancia estadistica.
    Igual Imprime cada una de ellas. 
    """
    #Lista Vacia    
    var_column = "igt:earn"
    #Rango por la cantidad total de preguntas que le pertenecen al NDC
    for j in range(1,21): 
        #Valor que toma por variable
        ndc = "beck:"+str(j)
        #Coeficiente de Pearson y P-valor entre las variables
        r, p_value = spearmanr(result[var_column], result[ndc])
        #print(r,p_value)
        if p_value <=p:
            string = f"El tiempo promedio con la pregunta #{str(j)} del ndc con un P-valor de: {p_value} y Coeficiente De Relaccion de: {r}" 
            lista.append(string)
            #lista2.append(r,p_value)
    return lista#,lista2

### P Valor < 0.05

In [28]:
lista3 = []
lista3 = igtevsbeck(lista3,p=0.05)
for element in lista3:
    print(Element)

### P Valor < 0.01

In [29]:
lista3 = []
lista3 = igtevsbeck(lista3,p=0.01)
for element in lista3:
    print(Element)

### P Valor < 0.001

In [30]:
lista3 = []
lista3 = igtevsbeck(lista3,p=0.001)
for element in lista3:
    print(Element)

## Respuestas


En este caso al realizar las correlaciones entre cada uno de las preguntas tanto del tiempo promedio de respuesta comparadas con el beck y el ndd, solo dos pasaron bajo una significancia de:
* P valor < 0.05 

Y ninguna paso bajo una significancia menor.
#### Tiempo Promedio vs BECK
Estas son las variables que mostraron una unica correlación:
* El tiempo promedio con la pregunta #1 del ndc con un P-valor de: 0.03232148859588864 y Coeficiente De Relaccion de: 0.43794581869247995
* El tiempo promedio con la pregunta #9 del ndc con un P-valor de: 0.036519567387405166 y Coeficiente De Relaccion de: -0.4288614750612719

En este caso, al haber comparado el dinero ganado con las respuestas del beck y del ndc ninguna mostro una existencia de correlación.

Debido a esto no ha habido necesidad de crear una tabla de representacion debido a la minima cantidad de variables que han aparecido.